clippy_lints/src/hashset_insert_after_contains.rs

```diff
@@ -0,0 +1,184 @@
+    /// Checks for usage of `contains` to see if a value is not
+    /// present on `HashSet` followed by a `insert`.
+    ///
+    /// ### Why is this bad?
+    /// Using just `insert` and checking the returned `bool` is more efficient.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::collections::HashSet;
+    /// let mut set = HashSet::new();
+    /// let value = 5;
+    /// if !set.contains(&value) {
+    ///     set.insert(value);
+    ///     println!("inserted {value:?}");
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::collections::HashSet;
+    /// let mut set = HashSet::new();
+    /// let value = 5;
+    /// if set.insert(&value) {
+    ///     println!("inserted {value:?}");
+    /// }
+    /// ```
+    #[clippy::version = "1.73.0"]
+    pub HASHSET_INSERT_AFTER_CONTAINS,
+    perf,
+    "use of `contains` to see if a value is not present on a `HashSet` followed by a `insert`"
+}
+declare_lint_pass!(HashsetInsertAfterContains => [HASHSET_INSERT_AFTER_CONTAINS]);
+
+impl<'tcx> LateLintPass<'tcx> for HashsetInsertAfterContains {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+
+        let Some(higher::If {
+            cond: cond_expr,
+            then: then_expr,
+            ..
+        }) = higher::If::hir(expr)
+        else {
+            return;
+        };
+
+        let Some(contains_expr) = try_parse_contains(cx, cond_expr) else {
+            return;
+        };
+
+        if !find_insert_calls(cx, &contains_expr, then_expr) {
+            return;
+        };
+        span_lint(
+            cx,
+            HASHSET_INSERT_AFTER_CONTAINS,
+            expr.span,
+            "usage of `HashSet::insert` after `HashSet::contains`. Remove the usage of `HashSet::contains` and just call `HashSet::contains` instead",
+        );
+    }
+}
+
+struct ContainsExpr<'tcx> {
+    receiver: &'tcx Expr<'tcx>,
+    value: &'tcx Expr<'tcx>,
+}
+fn try_parse_contains<'tcx>(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> Option<ContainsExpr<'tcx>> {
+    let expr = peel_hir_expr_while(expr, |e| match e.kind {
+        ExprKind::Unary(UnOp::Not, e) => Some(e),
+        _ => None,
+    });
+    match expr.kind {
+        ExprKind::MethodCall(
+            path,
+            receiver,
+            [
+                Expr {
+                    kind: ExprKind::AddrOf(_, _, value),
+                    span: value_span,
+                    ..
+                },
+            ],
+            _,
+        ) => {
+            let receiver_ty = cx.typeck_results().expr_ty(receiver);
+            if value_span.ctxt() == expr.span.ctxt()
+                && is_type_diagnostic_item(cx, receiver_ty, sym::HashSet)
+                && path.ident.name == sym!(contains)
+            {
+                Some(ContainsExpr { receiver, value })
+            } else {
+                None
+            }
```

Comment:
nit: Unnecessary `else`:
```suggestion
                return Some(ContainsExpr { receiver, value });
            }
            
            None
```